src/compute/src/compute_state.rs

```diff
@@ -937,6 +948,61 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
         // responses. This happens during shutdown.
         let _ = self.response_tx.send(response);
     }
+
+    /// Checks for dataflow expiration. Panics if we're past the replica expiration time.
+    pub(crate) fn check_expiration(&self) {
+        let now = mz_ore::now::SYSTEM_TIME();
+        if self.compute_state.replica_expiration.less_than(&now.into()) {
+            let now_datetime = mz_ore::now::to_datetime(now);
+            let expiration_datetime = self
+                .compute_state
+                .replica_expiration
+                .as_option()
+                .map(Into::into)
+                .map(mz_ore::now::to_datetime);
+
+            // We error and assert separately to produce structured logs in anything that depends
+            // on tracing.
+            error!(
+                now,
+                now_datetime = ?now_datetime,
+                expiration = ?self.compute_state.replica_expiration.elements(),
+                expiration_datetime = ?expiration_datetime,
+                "replica expired"
+            );
+
+            // Repeat condition for better error message.
+            assert!(
+                !self.compute_state.replica_expiration.less_than(&now.into()),
+                "replica expired. now: {now} ({now_datetime:?}), expiration: {:?} ({expiration_datetime:?})",
+                self.compute_state.replica_expiration.elements(),
+            );
+        }
+    }
+
+    /// Returns the dataflow expiration, i.e, the timestamp beyond which diffs can be
+    /// dropped.
+    ///
+    /// Returns an empty timestamp if `replica_expiration` is unset or matches conditions under
+    /// which dataflow expiration should be disabled.
```

Comment:
This comment talks about a returned timestamp, but I think it means "frontier"? At least the return type of the method is `Antichain` and I don't know what an empty timestamp is.